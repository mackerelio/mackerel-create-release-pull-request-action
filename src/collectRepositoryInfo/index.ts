import { ExecOptions } from "@actions/exec";
import * as github from "@actions/github";
import semver from "semver";

export type PullRequestInfo = {
  number: number;
  title: string;
  authorName: string;
  url: string;
};

// Get merged PRs between 'base' and 'head'
// https://github.com/Songmu/ghch/blob/c1a140de3b1cf5e65a895d28b45f36a2dca0a8b6/ghch.go#L242
export async function mergedPullRequests(params: {
  token: string;
  owner: string;
  repo: string;
  base: string;
  head: string;
}): Promise<PullRequestInfo[]> {
  const octokit = github.getOctokit(params.token);

  // Get commits between 'base' and 'head'
  const response = await octokit.rest.repos.compareCommits({
    owner: params.owner,
    repo: params.repo,
    base: params.base,
    head: params.head,
  });

  const results: PullRequestInfo[] = [];

  for (const { commit } of response.data.commits) {
    // Check if the commit is a merge commit
    const matched = /^Merge pull request #([0-9]+) from (\S+)/.exec(commit.message);
    if (!matched) {
      continue;
    }

    const prNumber = parseInt(matched[1], 10);
    const branchName = matched[2];
    // Request the PR detail using the PR number
    const { data } = await octokit.rest.pulls.get({
      owner: params.owner,
      repo: params.repo,
      pull_number: prNumber,
    });

    // https://github.com/Songmu/ghch/blob/c1a140de3b1cf5e65a895d28b45f36a2dca0a8b6/ghch.go#L265-L268
    if (data.head.label.replace(":", "/") !== branchName) {
      continue;
    }

    // Exclude nit PRs
    if (/\[nitp?\]/i.test(data.title)) {
      continue;
    }

    const authorName = data.user?.login ?? "";
    results.push({
      number: data.number,
      title: data.title,
      url: data.html_url,
      authorName,
    });
  }

  return results.sort((a, b) => b.number - a.number);
}

export async function getLatestReleaseVersion(params: {
  token: string;
  owner: string;
  repo: string;
}): Promise<string | null> {
  const octokit = github.getOctokit(params.token);
  const response = await octokit.rest.repos.getLatestRelease({
    owner: params.owner,
    repo: params.repo,
  });
  const tagName = response.data["tag_name"];
  return semver.valid(semver.coerce(tagName));
}

// use overridesOwnerAndRepo when debug and test.
export function getOwnerAndRepo(
  overridesOwnerAndRepo?: string
): Readonly<{ owner: string; repo: string }> {
  if (!overridesOwnerAndRepo) {
    return github.context.repo;
  }

  if (!overridesOwnerAndRepo.includes("/")) {
    throw new Error(
      `overridesOwnerAndRepo='${overridesOwnerAndRepo}' was not in a correct format; expected format {org}/{repo}.`
    );
  }

  const [owner, repo] = overridesOwnerAndRepo.split("/");
  return { owner, repo };
}

type CaptureCommandStdoutParams = Readonly<{
  exec: (commandLine: string, args?: string[], options?: ExecOptions) => Promise<number>;
  commandLine: string;
  args: string[] | undefined;
}>;

export async function captureCommandStdout(params: CaptureCommandStdoutParams): Promise<string> {
  let output = "";
  const options: ExecOptions = {};
  options.listeners = {
    stdout: data => {
      output += data.toString();
    },
  };

  await params.exec(params.commandLine, params.args, options);
  return output;
}

type GitDiffParams = Readonly<{
  exec: (commandLine: string, args?: string[], options?: ExecOptions) => Promise<number>;
  base: string;
  head: string;
  captureCommandStdout: (params: CaptureCommandStdoutParams) => Promise<string>;
}>;

export async function gitDiff(params: GitDiffParams): Promise<string> {
  return await params.captureCommandStdout({
    exec: params.exec,
    commandLine: "git",
    args: ["diff", "--word-diff", `${params.base}..${params.head}`],
  });
}
