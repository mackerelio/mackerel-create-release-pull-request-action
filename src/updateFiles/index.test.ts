import fs from "fs";
import { cp } from "@actions/io";
import snapshotDiff from "snapshot-diff";
import timezoneMock from "timezone-mock";
import { file } from "tmp-promise";
import { PullRequestInfo } from "../collectRepositoryInfo";
import {
  bumpMakefile,
  bumpVersionGo,
  updateDebPackageChangelog,
  updateRpmPackageChangelog,
  updateMarkdownChangelog,
  validatePackageName,
} from "./";

beforeAll(() => {
  timezoneMock.register("UTC");
});

afterAll(() => {
  timezoneMock.unregister();
});

describe("bumpVersionGo", () => {
  it("can rewrite version", async () => {
    const initialFile = "testdata/bumpVersionGo/version.go.txt";
    const initial = await fs.promises.readFile(initialFile, "utf8");

    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);

    await bumpVersionGo(testFilePath, "0.1.1");

    const processed = await fs.promises.readFile(testFilePath, "utf8");
    expect(snapshotDiff(initial, processed)).toMatchSnapshot();
  });
});

describe("bumpMakefile", () => {
  it("can rewrite version", async () => {
    const initialFile = "testdata/bumpMakefile/Makefile.txt";
    const initial = await fs.promises.readFile(initialFile, "utf8");

    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);

    await bumpMakefile(testFilePath, "0.1.2");

    const processed = await fs.promises.readFile(testFilePath, "utf8");
    expect(snapshotDiff(initial, processed)).toMatchSnapshot();
  });
});

const date = new Date(Date.UTC(2021, 1, 20, 7, 0, 0));

const prs: PullRequestInfo[] = [
  {
    title: "hoge huga",
    authorName: "piyo",
    url: "https://example.com/hogehuga",
    number: 3,
  },
  {
    title: "honyarara",
    authorName: "piyo",
    url: "https://example.com/honyarara",
    number: 4,
  },
];

describe("updateDebPackageChangelog", () => {
  test("can append text", async () => {
    const initialFile = "testdata/updateDebPackageChangelog/testing.txt";
    const initial = await fs.promises.readFile(initialFile, "utf8");

    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);

    await updateDebPackageChangelog(testFilePath, date, "test", "0.1.2", prs);

    const processed = await fs.promises.readFile(testFilePath, "utf8");
    expect(snapshotDiff(initial, processed)).toMatchSnapshot();
  });
  test("throw error when invalid package name", async () => {
    const initialFile = "testdata/updateDebPackageChangelog/testing.txt";
    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);
    await expect(
      updateDebPackageChangelog(testFilePath, date, "foo", "0.1.2", prs)
    ).rejects.toThrow("packageName is invalid. : foo");
  });
});

describe("updateRpmPackageChangelog", () => {
  test("can append text", async () => {
    const initialFile = "testdata/updateRpmPackageChangelog/testing.txt";
    const initial = await fs.promises.readFile(initialFile, "utf8");

    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);

    await updateRpmPackageChangelog(testFilePath, date, "0.1.2", prs);

    const processed = await fs.promises.readFile(testFilePath, "utf8");
    expect(snapshotDiff(initial, processed)).toMatchSnapshot();
  });
});

describe("updateMarkdownChangelog", () => {
  test("can append text", async () => {
    const initialFile = "testdata/updateMarkdownChangelog/testing.txt";
    const initial = await fs.promises.readFile(initialFile, "utf8");

    const { path: testFilePath } = await file();
    await cp(initialFile, testFilePath);

    await updateMarkdownChangelog(testFilePath, date, "0.1.2", prs);

    const processed = await fs.promises.readFile(testFilePath, "utf8");
    expect(snapshotDiff(initial, processed)).toMatchSnapshot();
  });
});

describe("validatePackageName", () => {
  test("can validate packageName", async () => {
    expect(() => {
      validatePackageName("mackerel-agent");
    }).not.toThrow();
  });
  test("throw exception when packageName is invalid", async () => {
    expect(() => {
      validatePackageName("invalid-packageName-**");
    }).toThrow("packageName is invalid");
  });
});
