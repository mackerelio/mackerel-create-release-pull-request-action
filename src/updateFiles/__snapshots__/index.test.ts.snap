// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bumpMakefile can rewrite version 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,7 @@
  BIN := mkr
- VERSION := 0.1.1
+ VERSION := 0.1.2
  CURRENT_REVISION := $(shell git rev-parse --short HEAD)
  BUILD_LDFLAGS := \\"-w -s -X main.gitcommit=$(CURRENT_REVISION)\\"

  export GO111MODULE=on
"
`;

exports[`bumpVersionGo can rewrite version 1`] = `
"Snapshot Diff:
- First value
+ Second value

  package main

- const version = \\"0.1.0\\"
+ const version = \\"0.1.1\\"

  var gitcommit string
"
`;

exports[`updateDebPackageChangelog can append text 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,5 +1,14 @@
+ test (0.1.2-1.v2) stable; urgency=low
+
+   * hoge huga (by piyo)
+     <https://example.com/hogehuga>
+   * honyarara (by piyo)
+     <https://example.com/honyarara>
+
+  -- mackerel <mackerel-developers@hatena.ne.jp>  Sat, 20 Feb 2021 07:00:00 +0000
+
  test (0.1.1-1.v2) stable; urgency=low

    * foo bar (by hoge)
      <https://example.com/foobar>
"
`;

exports[`updateMarkdownChangelog can append text 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,13 @@
  # Changelog

+ ## 0.1.2 (2021-02-20)
+
+ * hoge huga #3 (piyo)
+ * honyarara #4 (piyo)
+
+
  ## 0.1.1 (2021-02-18)

  * foo bar #2 (hoge)

"
`;

exports[`updateRpmPackageChangelog can append text 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,10 +1,14 @@
  %files
  %defattr(-,root,root)
  %{_bindir}/%{name}

  %changelog
+ * Sat Feb 20 2021 <mackerel-developers@hatena.ne.jp> - 0.1.2
+ - hoge huga (by piyo)
+ - honyarara (by piyo)
+
  * Thu Feb 18 2021 <mackerel-developers@hatena.ne.jp> - 0.1.1
  - foo bar (by hoge)

  * Thu Feb 16 2021 <mackerel-developers@hatena.ne.jp> - 0.1.0
  - baz qux (by huga)"
`;
